lab3van/util-linux/schedutils/taskset.1
.\" taskset(1) manpage
.\"
.\" Copyright (C) 2004 Robert Love
.\"
.\" This is free documentation; you can redistribute it and/or
<<<<<<< HEAD
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License.
=======
.\" modify it under the terms of the GNU General Public License,
.\" version 2, as published by the Free Software Foundation.
>>>>>>> master-vanilla
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
<<<<<<< HEAD
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.\" 2002-05-11 Robert Love <rml@tech9.net>
.\" 	Initial version
.\"
.TH TASKSET 1 "April 2003" "util-linux" "User Commands"
.SH NAME
taskset \- retrieve or set a process's CPU affinity
.SH SYNOPSIS
.B taskset
.RI [ options ]\  mask
.IR command\  [ arg ]...
.br
.B taskset
.RI [ options ]
=======
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH TASKSET 1 "August 2014" "util-linux" "User Commands"
.SH NAME
taskset \- set or retrieve a process's CPU affinity
.SH SYNOPSIS
.B taskset
[options]
.IR mask\ command\  [ argument ...]
.br
.B taskset
[options]
>>>>>>> master-vanilla
.B \-p
.RI [ mask ]\  pid
.SH DESCRIPTION
.PP
.B taskset
<<<<<<< HEAD
is used to set or retrieve the CPU affinity of a running process given its PID
or to launch a new COMMAND with a given CPU affinity.  CPU affinity is a
=======
is used to set or retrieve the CPU affinity of a running process given its
\fIpid\fR, or to launch a new \fIcommand\fR with a given CPU affinity.
CPU affinity is a
>>>>>>> master-vanilla
scheduler property that "bonds" a process to a given set of CPUs on the system.
The Linux scheduler will honor the given CPU affinity and the process will not
run on any other CPUs.  Note that the Linux scheduler also supports natural
CPU affinity: the scheduler attempts to keep processes on the same CPU as long
as practical for performance reasons.  Therefore, forcing a specific CPU
affinity is useful only in certain applications.
.sp
The CPU affinity is represented as a bitmask, with the lowest order bit
corresponding to the first logical CPU and the highest order bit corresponding
to the last logical CPU.  Not all CPUs may exist on a given system but a mask
may specify more CPUs than are present.  A retrieved mask will reflect only the
bits that correspond to CPUs physically on the system.  If an invalid mask is
given (i.e., one that corresponds to no valid CPUs on the current system) an
error is returned.  The masks are typically given in hexadecimal.  For example,
<<<<<<< HEAD
.TP
.BR 0x00000001
is processor #0
.TP
.BR 0x00000003
is processors #0 and #1
.TP
.BR 0xFFFFFFFF
is all processors (#0 through #31).
=======
.RS 4
.TP 12
.BR 0x00000001
is processor #0,
.TP
.BR 0x00000003
is processors #0 and #1,
.TP
.BR 0xFFFFFFFF
is all processors (#0 through #31).
.RE
>>>>>>> master-vanilla
.PP
When
.BR taskset
returns, it is guaranteed that the given program has been scheduled to a legal
CPU.
.SH OPTIONS
.TP
<<<<<<< HEAD
.BR \-a ,\  \-\-all-tasks
Set or retrieve the CPU affinity of all the tasks (threads) for a given PID.
.TP
.BR \-p ,\  \-\-pid
Operate on an existing PID and do not launch a new task.
.TP
.BR \-c ,\  \-\-cpu-list
Specify a numerical list of processors instead of a bitmask.  The numbers
are separated by commas and may include ranges.  For example:
.BR 0,5,7,9-11 .
.TP
.BR \-h ,\  \-\-help
Display usage information and exit.
.TP
.BR \-V ,\  \-\-version
Display version information and exit.
=======
.BR \-a ,\  \-\-all\-tasks
Set or retrieve the CPU affinity of all the tasks (threads) for a given PID.
.TP
.BR \-c ,\  \-\-cpu\-list \ \fInumbers
Specify a numerical list of processors instead of a bitmask.  The \fInumbers\fR
are separated by commas and may include ranges.  For example:
.BR 0,5,8-11 .
.TP
.BR \-p ,\  \-\-pid
Operate on an existing PID and do not launch a new task.
.TP
.BR \-V ,\  \-\-version
Display version information and exit.
.TP
.BR \-h ,\  \-\-help
Display help text and exit.
>>>>>>> master-vanilla
.SH USAGE
.TP
The default behavior is to run a new command with a given affinity mask:
.B taskset
.I mask
.IR command\  [ arguments ]
.TP
You can also retrieve the CPU affinity of an existing task:
.B taskset \-p
.I pid
.TP
Or set it:
.B taskset \-p
.I mask pid
.SH PERMISSIONS
<<<<<<< HEAD
A user must possess
.B CAP_SYS_NICE
to change the CPU affinity of a process.  Any user can retrieve the affinity
mask.
.SH AUTHOR
Written by Robert M. Love.
.SH COPYRIGHT
Copyright \(co 2004 Robert M. Love
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
=======
A user can change the CPU affinity of a process belonging to the same user.
A user must possess
.B CAP_SYS_NICE
to change the CPU affinity of a process belonging to another user.
A user can retrieve the affinity mask of any process.
.SH SEE ALSO
>>>>>>> master-vanilla
.BR chrt (1),
.BR nice (1),
.BR renice (1),
.BR sched_setaffinity (2),
.BR sched_getaffinity (2)
.sp
See
.BR sched_setscheduler (2)
for a description of the Linux scheduling scheme.
<<<<<<< HEAD
=======
.SH AUTHOR
Written by Robert M. Love.
.SH COPYRIGHT
Copyright \(co 2004 Robert M. Love.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
>>>>>>> master-vanilla
.SH AVAILABILITY
The taskset command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/schedutils/taskset.c
/*
<<<<<<< HEAD
 * taskset.c - command-line utility for setting and retrieving
 *             a task's CPU affinity
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, v2, as
=======
 * taskset.c - set or retrieve a task's CPU affinity
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
>>>>>>> master-vanilla
 * published by the Free Software Foundation
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
<<<<<<< HEAD
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
=======
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
>>>>>>> master-vanilla
 *
 * Copyright (C) 2004 Robert Love
 * Copyright (C) 2010 Karel Zak <kzak@redhat.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <sched.h>
#include <stddef.h>
#include <string.h>

#include "cpuset.h"
#include "nls.h"
#include "strutils.h"
#include "xalloc.h"
#include "procutils.h"
#include "c.h"
<<<<<<< HEAD
=======
#include "closestream.h"
>>>>>>> master-vanilla

struct taskset {
	pid_t		pid;		/* task PID */
	cpu_set_t	*set;		/* task CPU mask */
	size_t		setsize;
	char		*buf;		/* buffer for conversion from mask to string */
	size_t		buflen;
	unsigned int	use_list:1,	/* use list rather than masks */
			get_only:1;	/* print the mask, but not modify */
};

static void __attribute__((__noreturn__)) usage(FILE *out)
{
	fprintf(out,
		_("Usage: %s [options] [mask | cpu-list] [pid|cmd [args...]]\n\n"),
		program_invocation_short_name);

<<<<<<< HEAD
=======
	fputs(USAGE_SEPARATOR, out);
	fputs(_("Show or change the CPU affinity of a process.\n"), out);
	fputs(USAGE_SEPARATOR, out);

>>>>>>> master-vanilla
	fprintf(out, _(
		"Options:\n"
		" -a, --all-tasks         operate on all the tasks (threads) for a given pid\n"
		" -p, --pid               operate on existing given pid\n"
		" -c, --cpu-list          display and specify cpus in list format\n"
		" -h, --help              display this help\n"
		" -V, --version           output version information\n\n"));

	fprintf(out, _(
		"The default behavior is to run a new command:\n"
		"    %1$s 03 sshd -b 1024\n"
		"You can retrieve the mask of an existing task:\n"
		"    %1$s -p 700\n"
		"Or set it:\n"
		"    %1$s -p 03 700\n"
		"List format uses a comma-separated list instead of a mask:\n"
		"    %1$s -pc 0,3,7-11 700\n"
		"Ranges in list format can take a stride argument:\n"
		"    e.g. 0-31:2 is equivalent to mask 0x55555555\n"),
		program_invocation_short_name);

<<<<<<< HEAD
	fprintf(out, _("\nFor more information see taskset(1).\n"));
=======
	fprintf(out, USAGE_MAN_TAIL("taskset(1)"));
>>>>>>> master-vanilla

	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

static void print_affinity(struct taskset *ts, int isnew)
{
	char *str, *msg;

	if (ts->use_list) {
		str = cpulist_create(ts->buf, ts->buflen, ts->set, ts->setsize);
		msg = isnew ? _("pid %d's new affinity list: %s\n") :
			      _("pid %d's current affinity list: %s\n");
	} else {
		str = cpumask_create(ts->buf, ts->buflen, ts->set, ts->setsize);
		msg = isnew ? _("pid %d's new affinity mask: %s\n") :
			      _("pid %d's current affinity mask: %s\n");
	}

	if (!str)
<<<<<<< HEAD
		/* this is internal error... */
		errx(EXIT_FAILURE, _("conversion from cpuset to string failed"));
=======
		errx(EXIT_FAILURE, _("internal error: conversion from cpuset to string failed"));
>>>>>>> master-vanilla

	printf(msg, ts->pid, str);
}

static void do_taskset(struct taskset *ts, size_t setsize, cpu_set_t *set)
{
	/* read the current mask */
	if (ts->pid) {
		if (sched_getaffinity(ts->pid, ts->setsize, ts->set) < 0)
			err(EXIT_FAILURE, _("failed to get pid %d's affinity"),
			    ts->pid);
		print_affinity(ts, FALSE);
	}

	if (ts->get_only)
		return;

	/* set new mask */
	if (sched_setaffinity(ts->pid, setsize, set) < 0)
		err(EXIT_FAILURE, _("failed to set pid %d's affinity"),
		    ts->pid);

	/* re-read the current mask */
	if (ts->pid) {
		if (sched_getaffinity(ts->pid, ts->setsize, ts->set) < 0)
			err(EXIT_FAILURE, _("failed to get pid %d's affinity"),
			    ts->pid);
		print_affinity(ts, TRUE);
	}
}

int main(int argc, char **argv)
{
	cpu_set_t *new_set;
	pid_t pid = 0;
	int c, all_tasks = 0;
<<<<<<< HEAD
	unsigned int ncpus;
=======
	int ncpus;
>>>>>>> master-vanilla
	size_t new_setsize, nbits;
	struct taskset ts;

	static const struct option longopts[] = {
		{ "all-tasks",	0, NULL, 'a' },
		{ "pid",	0, NULL, 'p' },
		{ "cpu-list",	0, NULL, 'c' },
		{ "help",	0, NULL, 'h' },
		{ "version",	0, NULL, 'V' },
		{ NULL,		0, NULL,  0  }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
<<<<<<< HEAD
=======
	atexit(close_stdout);
>>>>>>> master-vanilla

	memset(&ts, 0, sizeof(ts));

	while ((c = getopt_long(argc, argv, "+apchV", longopts, NULL)) != -1) {
		switch (c) {
		case 'a':
			all_tasks = 1;
			break;
		case 'p':
<<<<<<< HEAD
			pid = strtol_or_err(argv[argc - 1],
					    _("failed to parse pid"));
=======
			pid = strtos32_or_err(argv[argc - 1],
					    _("invalid PID argument"));
>>>>>>> master-vanilla
			break;
		case 'c':
			ts.use_list = 1;
			break;
		case 'V':
<<<<<<< HEAD
			printf("%s from %s\n", program_invocation_short_name,
			       PACKAGE_STRING);
=======
			printf(UTIL_LINUX_VERSION);
>>>>>>> master-vanilla
			return EXIT_SUCCESS;
		case 'h':
			usage(stdout);
			break;
		default:
			usage(stderr);
			break;
		}
	}

	if ((!pid && argc - optind < 2)
	    || (pid && (argc - optind < 1 || argc - optind > 2)))
		usage(stderr);

	ncpus = get_max_number_of_cpus();
	if (ncpus <= 0)
		errx(EXIT_FAILURE, _("cannot determine NR_CPUS; aborting"));

	/*
	 * the ts->set is always used for the sched_getaffinity call
	 * On the sched_getaffinity the kernel demands a user mask of
	 * at least the size of its own cpumask_t.
	 */
	ts.set = cpuset_alloc(ncpus, &ts.setsize, &nbits);
	if (!ts.set)
		err(EXIT_FAILURE, _("cpuset_alloc failed"));

	/* buffer for conversion from mask to string */
	ts.buflen = 7 * nbits;
	ts.buf = xmalloc(ts.buflen);

	/*
	 * new_set is always used for the sched_setaffinity call
	 * On the sched_setaffinity the kernel will zero-fill its
	 * cpumask_t if the user's mask is shorter.
	 */
	new_set = cpuset_alloc(ncpus, &new_setsize, NULL);
	if (!new_set)
		err(EXIT_FAILURE, _("cpuset_alloc failed"));

	if (argc - optind == 1)
		ts.get_only = 1;

	else if (ts.use_list) {
<<<<<<< HEAD
		if (cpulist_parse(argv[optind], new_set, new_setsize))
=======
		if (cpulist_parse(argv[optind], new_set, new_setsize, 0))
>>>>>>> master-vanilla
			errx(EXIT_FAILURE, _("failed to parse CPU list: %s"),
			     argv[optind]);
	} else if (cpumask_parse(argv[optind], new_set, new_setsize)) {
		errx(EXIT_FAILURE, _("failed to parse CPU mask: %s"),
		     argv[optind]);
	}

<<<<<<< HEAD
	if (all_tasks) {
=======
	if (all_tasks && pid) {
>>>>>>> master-vanilla
		struct proc_tasks *tasks = proc_open_tasks(pid);
		while (!proc_next_tid(tasks, &ts.pid))
			do_taskset(&ts, new_setsize, new_set);
		proc_close_tasks(tasks);
	} else {
		ts.pid = pid;
		do_taskset(&ts, new_setsize, new_set);
	}

	free(ts.buf);
	cpuset_free(ts.set);
	cpuset_free(new_set);

	if (!pid) {
		argv += optind + 1;
		execvp(argv[0], argv);
<<<<<<< HEAD
		err(EXIT_FAILURE, _("executing %s failed"), argv[0]);
=======
		err(EXIT_FAILURE, _("failed to execute %s"), argv[0]);
>>>>>>> master-vanilla
	}

	return EXIT_SUCCESS;
}
